### DISCUSSION TITLE
ADC 12bit auto accumulation for oversampling with Arduino on 1624 issue

### DISCUSSION THREAD

**Discussion by DietmarM** [SpenceKonde/megaTinyCore#554] [Q&A] [answered]
I try to use the new ADC 12 bit with accumulation - but struggle to get the ADC working at all. I get readings when I run the Serial.print(analogRead(A5));  command further down in the code - but only 0 if I don't. But as I would want to use the enhanced oversampling paralell to the loop this is not the way to read analog data.

ADC0.RESULT is allways 0 if analogRead is not called

This is basicly the blink example but added ADC reading on PA5 which is Pin3 on my 1624.

#define LED_BUILTIN 0 //PA4 (Pin 2)

#define TIMEBASE_VALUE (uint8_t)ceil(F_CPU*0.000002) //2µs Abstand zw. ADC Messungen --> 100Werte sollen 0,2ms dauern?
#define ADC_MAX_VALUE ((1 << 12) - 1) /* In single-ended mode, the max value is 4095 */
/* Defines to easily configure ADC accumulation */
#define ADC_SAMPNUM_CONFIG ADC_SAMPNUM_ACC256_gc
/* Left shifting (1 << SAMPNUM) results in the number of accumulated samples */
#define ADC_SAMPLES (1 << ADC_SAMPNUM_CONFIG)

/* Volatile variables to improve debug experience */
static volatile uint32_t adc_reading;
static volatile float voltage;

// the setup function runs once when you press reset or power the board
void setup() {
  // initialize digital pin LED_BUILTIN as an output.
  pinMode(LED_BUILTIN, OUTPUT);
  //pinMode(PIN_PA5, INPUT);
  //PORTA.PIN5CTRL |= PORT_ISC_INPUT_DISABLE_gc; //PA5 analog
  
    //Serial.begin(9600); // Setup der seriellen Verbindung
  Serial.begin(115200); // Setup der seriellen Verbindung

  PORTA.PIN5CTRL |= PORT_ISC_INPUT_DISABLE_gc; //PA5 analog

  ADC0.CTRLA = ADC_ENABLE_bm;
  ADC0.CTRLB = ADC_PRESC_DIV4_gc; /*MAX FCLK_ADC 2.5 Mhz*/
  ADC0.CTRLC = ADC_REFSEL_VDD_gc | TIMEBASE_VALUE << ADC_TIMEBASE0_bp;/*ADC reference VDD*/
  
  ADC0.CTRLE = 3; /*SAMPDUR*/
  ADC0.CTRLF = ADC_FREERUN_bm;      /*Single*/
  ADC0.MUXPOS = ADC_MUXPOS_AIN5_gc;   /*ADC channel AIN5->PA5*/
  ADC0.INTFLAGS = ADC_RESRDY_bm;
  ADC0.COMMAND = ADC_MODE_SINGLE_12BIT_gc; /*Single*/
    
}

// the loop function runs over and over again forever
void loop() {
  digitalWrite(LED_BUILTIN, HIGH);   // turn the LED on (HIGH is the voltage level)
  delay(29);

  ADC0.COMMAND = ADC_MODE_SINGLE_12BIT_gc; 

  //Serial.print(analogRead(A5));
  Serial.print(" ");
  Serial.print(ADC0.SAMPLE);

  if(ADC0.INTFLAGS & ADC_RESRDY_bm)
  {
     adc_reading = ADC0.RESULT; /* Read ADC result */
     Serial.print("JA: ");
  }

  
  
 /* Calculate voltage on ADC pin, VDD = 3.3V, 12-bit resolution, 256 samples */
   
  Serial.print(adc_reading);
  Serial.print("ADC ");

  Serial.print("Hello"); // Wert ausgeben
  
  delay(1000);                       // wait for a second
  digitalWrite(LED_BUILTIN, LOW);    // turn the LED off by making the voltage LOW
  delay(200);                       // wait for a second
  Serial.println("bye"); // Wert ausgeben
}

**Comment by SpenceKonde** [comment, marked as answer]
Hae you tried like, just using the API the core provides for doing that, analogReadEnh()? (or analogReadDiff() for differential? It lets you specify raw data or totals, and you can also have it automatically decimate it for you? 

https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/Ref_Analog.md

**Reply by DietmarM** [reply]
This looks interesting. How can I start the ADC to do addup of e.g. 250 readings and take the result later?
Is the function analogReadDiff(PIN_PA1, PIN_PA2, 12, 0);
doing the readings and the rest of the program is not executed?
I would like to utilise the new ADC as much as possible to get the best reading I can get...

**Comment by SpenceKonde** [comment]
I would suggest confirming that that works, and then trying to implement it yourself if need be.

**Reply by DietmarM** [reply]
Did some tests and the Enhanced functions work as expected

**Comment by SpenceKonde** [comment]
The analogReadEnh and analogReadDiff() functions are blocking - like analogRead; it sound likeyou want either triggered readings or freerunning mode. I know that they have free running mode, I remember looking over it and not playing with it because it was less important that other tasks on my list. It definitely should allow you to take readings in the background. 

You can also set it so it will take readings in reaction to an event input through the EVSYS, which can be controlled by timers, pins, almost anything, including logical combinations of them through the CCL. It;'sa very powerful system for sure. 


Anyway, in free running mode , the most recent measurement is available at the SAMPLE register (soyou can monitor the voltage as the measurement proceeds, and at the end, the full result bill be in the RESULT register (may be up to 17 bits in size, so read it into a long if doing more than 512 accumulations, 
You'd do all the things that my analogReadEnh does to set it up, except that you set the FreeRunning bit, and either enable the ADC result ready interrupt, or poll ADC0.STATUS periodically depending on your needs, and then go do something else. analogReadEnh() has the behavior of analogRead() - it waits for a sample and then returns a value. 

If it was getting things in the background, how would you tell it what to do with those values? Don't say attach interruot, I've spent the past week rewriting things that call functions in ISRs in hand tuned assembly because if the brutal overhead
I would say ou;d want to mostly configure that manually anyway

**Reply by DietmarM** [reply]
Thanks so much for your answer - this sound promissing to me.
analogReadEnh: May I ask where to find the source code?
